### PR TITLE
[Mobile] Update build_mobile.sh to allow lite interpreter and tracing based builds

### PR DESCRIPTION
**Pull request by dhruvbird**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84704
* __->__ #84647

Summary: Currently, build_mobile.sh doesn't allow lite interpreter builds or tracing based selective builds. build_mobile.sh is used for host builds of PyTorch for Mobile deployment.

Additionally, certain flags such as `USE_BLAS` were not being respected as they should be. This change addresses that as well.

Test Plan: Build using:

```
cat /tmp/selected_ops.yaml
- aten::add
- aten::sub
```

```
BUILD_PYTORCH_MOBILE_WITH_HOST_TOOLCHAIN=1 USE_LIGHTWEIGHT_DISPATCH=0 BUILD_LITE_INTERPRETER=1 SELECTED_OP_LIST=/tmp/selected_ops.yaml ./scripts/build_mobile.sh
```

```
cat /tmp/main.cpp

int main() {
  auto m = torch::jit::_load_for_mobile("/tmp/path_to_model.ptl");
  auto res = m.forward({});
  return 0;
}
```

Test using:

```
g++ /tmp/main.cpp -L build_mobile/lib/ -I build_mobile/install/include/ -lpthread -lc10 -ltorch_cpu -ltorch -lXNNPACK -lpytorch_qnnpack -lcpuinfo -lclog -lpthreadpool -lgloo -lkineto -lfmt -ldl -lc10
```

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Comment by cccclai** [comment]
This script seems to be used for testing. I usually use `./scripts/build_ios.sh` or `./scripts/build_pytorch_android.sh`. Any specific reason to update this script?

**Comment by dhruvbird** [comment]
> This script seems to be used for testing. I usually use `./scripts/build_ios.sh` or `./scripts/build_pytorch_android.sh`. Any specific reason to update this script?

`build_mobile.sh` is really useful when developing PyTorch on a Linux machine to check the sanity of the build, etc... Keeping this updated with the Android and iOS build will help developer productivity and allow folks to ensure that the x-platform Linux (host) build works before they try the Android/iOS specific builds.

**Comment by dhruvbird** [comment]
The failure for [linux-bionic-cuda11.6-py3.10-gcc7 / test (default, 1, 4, linux.4xlarge.nvidia.gpu)](https://github.com/pytorch/pytorch/runs/8239399837?check_suite_focus=true#logs) is unrelated to this change. It reads:

> RuntimeError: CUDA driver error: out of memory

**Review by JacobSzwejbka — APPROVED** [review]

File: scripts/build_mobile.sh [hunk 1]
```diff
@@ -35,6 +36,32 @@ if [ -x "$(command -v ninja)" ]; then
...
+
+# If there exists env variable and it equals to 1, build lite interpreter.
+# Default behavior is to build full jit interpreter.
+# cmd:  BUILD_LITE_INTERPRETER=1 ./scripts/build_mobile.sh
+if [ "x${BUILD_LITE_INTERPRETER}" == "x1" ]; then
+  CMAKE_ARGS+=("-DBUILD_LITE_INTERPRETER=ON")
+else
+  CMAKE_ARGS+=("-DBUILD_LITE_INTERPRETER=OFF")
+fi
+if [ "x${TRACING_BASED}" == "x1" ]; then
```
Comment (cccclai) [inline scripts/build_mobile.sh:51]: Why is there an "x" here?

**Review by cccclai — APPROVED** [review]
LGTM. Thanks!

File: scripts/build_mobile.sh (hunk 1 above)
Comment (dhruvbird) [reply scripts/build_mobile.sh:51]: Good question!

https://stackoverflow.com/questions/3869072/test-for-non-zero-length-string-in-bash-n-var-or-var/49825114#49825114

I just learned that:

> This is very old-school from the Bourne shell days. Don't perpetuate this old habit. bash is a sharper tool than its predecessors.

**Comment by dhruvbird** [comment]
https://github.com/pytorch/pytorch/pull/84704 is stacked on top of this change, and all tests pass in CI, so I'm going to force merge this one and then the next one (in that order).

@pytorchbot merge -f "CI pass on the next PR in this stack, and the 1 failure here seems unrelated to this change - out of memory error"

**Pull request closed without merging**